File: app/src/main/java/com/codylimber/fieldphenology/ui/theme/AppSettings.kt
package com.codylimber.fieldphenology.ui.theme

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import android.content.SharedPreferences
import com.codylimber.fieldphenology.data.api.TaxonResult
import com.codylimber.fieldphenology.data.model.SortMode
import org.json.JSONArray
import org.json.JSONObject

object AppSettings {
    var useScientificNames by mutableStateOf(false)
    var minActivityPercent by mutableStateOf(0)
    var defaultSortMode by mutableStateOf(SortMode.LIKELIHOOD)
    var favorites by mutableStateOf(setOf<Int>())
    // Favorite taxa the user wants to quickly load when building a new dataset
    // (e.g. Butterflies, a few plant genera). Distinct from `favorites`, which
    // is starred individual species within datasets.
    var favoriteTaxa by mutableStateOf(listOf<TaxonResult>())
    var selectedDatasetKeys by mutableStateOf(setOf<String>())
    var showActiveOnly by mutableStateOf(true)
    var targetMode by mutableStateOf("STARRED")
    var weeklyDigestEnabled by mutableStateOf(false)
    var digestDays by mutableStateOf(setOf(2)) // Calendar day constants: Sunday=1..Saturday=7
    var digestHour by mutableStateOf(8)
    var targetNotificationsEnabled by mutableStateOf(false)
    var notificationDatasetKeys by mutableStateOf(setOf<String>()) // empty = all datasets

    private var prefs: SharedPreferences? = null

    fun init(prefs: SharedPreferences) {
        this.prefs = prefs
        favorites = prefs.getStringSet("favorites", emptySet())!!.mapNotNull { it.toIntOrNull() }.toSet()
        favoriteTaxa = parseFavoriteTaxa(prefs.getString("favorite_taxa", null))
        weeklyDigestEnabled = prefs.getBoolean("weekly_digest_enabled", false)
        // Migrate old single digestDay to digestDays set
        val oldDay = prefs.getInt("digest_day", -1)
        digestDays = prefs.getStringSet("digest_days", null)?.mapNotNull { it.toIntOrNull() }?.toSet()
            ?: if (oldDay >= 1) setOf(oldDay) else setOf(2)
        digestHour = prefs.getInt("digest_hour", 8)
        targetNotificationsEnabled = prefs.getBoolean("target_notifications_enabled", false)
        notificationDatasetKeys = prefs.getStringSet("notification_dataset_keys", emptySet()) ?: emptySet()
        selectedDatasetKeys = prefs.getStringSet("selected_dataset_keys", emptySet()) ?: emptySet()
        showActiveOnly = prefs.getBoolean("show_active_only", true)
        targetMode = prefs.getString("target_mode", "STARRED") ?: "STARRED"
    }

    fun saveSelectedDatasetKeys() {
        prefs?.edit()?.putStringSet("selected_dataset_keys", selectedDatasetKeys)?.apply()
    }

    fun saveShowActiveOnly() {
        prefs?.edit()?.putBoolean("show_active_only", showActiveOnly)?.apply()
    }

    fun saveTargetMode() {
        prefs?.edit()?.putString("target_mode", targetMode)?.apply()
    }

    fun toggleFavorite(taxonId: Int) {
        favorites = if (taxonId in favorites) favorites - taxonId else favorites + taxonId
        prefs?.edit()?.putStringSet("favorites", favorites.map { it.toString() }.toSet())?.apply()
    }

    fun isFavorite(taxonId: Int): Boolean = taxonId in favorites

    fun addFavoriteTaxon(taxon: TaxonResult) {
        if (favoriteTaxa.none { it.id == taxon.id }) {
            favoriteTaxa = favoriteTaxa + taxon
            saveFavoriteTaxa()
        }
    }

    fun removeFavoriteTaxon(taxonId: Int) {
        favoriteTaxa = favoriteTaxa.filter { it.id != taxonId }
        saveFavoriteTaxa()
    }

    private fun saveFavoriteTaxa() {
        val arr = JSONArray()
        favoriteTaxa.forEach { t ->
            arr.put(JSONObject().apply {
                put("id", t.id)
                put("displayName", t.displayName)
                put("scientificName", t.scientificName)
                put("commonName", t.commonName)
                put("rank", t.rank)
            })
        }
        prefs?.edit()?.putString("favorite_taxa", arr.toString())?.apply()
    }

    private fun parseFavoriteTaxa(json: String?): List<TaxonResult> {
        if (json.isNullOrBlank()) return emptyList()
        return try {
            val arr = JSONArray(json)
            (0 until arr.length()).map { i ->
                val o = arr.getJSONObject(i)
                TaxonResult(
                    id = o.getInt("id"),
                    displayName = o.optString("displayName"),
                    scientificName = o.optString("scientificName"),
                    commonName = o.optString("commonName"),
                    rank = o.optString("rank")
                )
            }
        } catch (_: Exception) {
            emptyList()
        }
    }
}

File: ios/Manakin/UI/Theme/AppSettings.swift
import Foundation

@Observable
final class AppSettings {
    static let shared = AppSettings()

    var useScientificNames: Bool {
        didSet { defaults.set(useScientificNames, forKey: "useScientificNames") }
    }
    var minActivityPercent: Int {
        didSet { defaults.set(minActivityPercent, forKey: "minActivityPercent") }
    }
    var defaultSortMode: SortMode {
        didSet { defaults.set(defaultSortMode.rawValue, forKey: "defaultSortMode") }
    }
    var favorites: Set<Int> {
        didSet { defaults.set(Array(favorites), forKey: "favorites") }
    }
    var selectedDatasetKeys: Set<String> {
        didSet { defaults.set(Array(selectedDatasetKeys), forKey: "selectedDatasetKeys") }
    }
    var showActiveOnly: Bool {
        didSet { defaults.set(showActiveOnly, forKey: "showActiveOnly") }
    }
    var targetMode: String {
        didSet { defaults.set(targetMode, forKey: "targetMode") }
    }
    var weeklyDigestEnabled: Bool {
        didSet { defaults.set(weeklyDigestEnabled, forKey: "weeklyDigestEnabled") }
    }
    var digestDays: Set<Int> {
        didSet { defaults.set(Array(digestDays), forKey: "digestDays") }
    }
    var digestHour: Int {
        didSet { defaults.set(digestHour, forKey: "digestHour") }
    }
    var targetNotificationsEnabled: Bool {
        didSet { defaults.set(targetNotificationsEnabled, forKey: "targetNotificationsEnabled") }
    }
    var notificationDatasetKeys: Set<String> {
        didSet { defaults.set(Array(notificationDatasetKeys), forKey: "notificationDatasetKeys") }
    }

    private let defaults = UserDefaults.standard

    private init() {
        self.useScientificNames = defaults.bool(forKey: "useScientificNames")
        self.minActivityPercent = defaults.object(forKey: "minActivityPercent") as? Int ?? 0
        let sortRaw = defaults.string(forKey: "defaultSortMode") ?? "likelihood"
        self.defaultSortMode = SortMode(rawValue: sortRaw) ?? .likelihood
        let favArray = defaults.array(forKey: "favorites") as? [Int] ?? []
        self.favorites = Set(favArray)
        let dsKeys = defaults.array(forKey: "selectedDatasetKeys") as? [String] ?? []
        self.selectedDatasetKeys = Set(dsKeys)
        self.showActiveOnly = defaults.object(forKey: "showActiveOnly") as? Bool ?? true
        self.targetMode = defaults.string(forKey: "targetMode") ?? "STARRED"
        self.weeklyDigestEnabled = defaults.bool(forKey: "weeklyDigestEnabled")
        let daysArray = defaults.array(forKey: "digestDays") as? [Int] ?? [2]
        self.digestDays = Set(daysArray)
        self.digestHour = defaults.object(forKey: "digestHour") as? Int ?? 8
        self.targetNotificationsEnabled = defaults.bool(forKey: "targetNotificationsEnabled")
        let notifKeys = defaults.array(forKey: "notificationDatasetKeys") as? [String] ?? []
        self.notificationDatasetKeys = Set(notifKeys)
    }

    func toggleFavorite(_ taxonId: Int) {
        if favorites.contains(taxonId) {
            favorites.remove(taxonId)
        } else {
            favorites.insert(taxonId)
        }
    }

    func isFavorite(_ taxonId: Int) -> Bool {
        favorites.contains(taxonId)
    }
}

extension SortMode: RawRepresentable {
    public init?(rawValue: String) {
        switch rawValue {
        case "likelihood": self = .likelihood
        case "peakDate": self = .peakDate
        case "name": self = .name
        case "taxonomy": self = .taxonomy
        default: return nil
        }
    }

    public var rawValue: String {
        switch self {
        case .likelihood: return "likelihood"
        case .peakDate: return "peakDate"
        case .name: return "name"
        case .taxonomy: return "taxonomy"
        }
    }
}
